### PR TITLE
feat(telemetry): add gateway info to reported data

### PR DESCRIPTION
**Pull request by thalesmg**
Example data:

```json
"gateway": {
    "lwm2m": {
      "num_clients": 0,
      "authn": "password-based:mysql",
      "listeners": [
        {
          "type": "tcp", // tcp | ssl | udp | dtls
          "authn": "password-based:mysql"
        }
      ]
    }
  },
```

### CODE REVIEW

**Review by savonarola — DISMISSED** [review]

File: apps/emqx_gateway/src/emqx_gateway.erl [hunk 1]
```diff
@@ -81,6 +84,66 @@ start(Name) ->
...
+        end,
+        #{},
+        list()
+    ).
+
 %%--------------------------------------------------------------------
 %% Internal funcs
 %%--------------------------------------------------------------------
+
+get_authn_type(#{authentication := #{mechanism := Mechanism, backend := Backend}}) when
```
Comment (zmstone) [inline apps/emqx_gateway/src/emqx_gateway.erl:126]: call emqx_authentication_config:authn_type/1 ?

Comment (zmstone) [reply apps/emqx_gateway/src/emqx_gateway.erl:126]: actually `emqx_authentication_config:authenticator_id/1`

**Comment by thalesmg** [comment]
> M.b. this isn't important, but we seem to introduce implicit circular dependencies between apps: `emqx_telemetry` calls functions from other apps explicitly; at the same time we have funs in other apps specially constructed for telemetry, so they "know" about telemetry too.
> 
> To get rid of this, we may for example introduce some IoC, like `minirest` does:
> 
>     * introduce some `emqx_telemetry` behaviour;
> 
>     * implement it in apps that need to send data to telemetry
> 
>     * in telemetry app discover modules implementing  `emqx_telemetry` behaviour and gather metrics from them.
>       This will make dependencies unidirectional: apps will depend on an abstract interface.
> 
> 
> Maybe this isn't worth doing :)

That's quite an interesting idea!  Thanks!

I wonder if that would change things a lot, in particular because each application outputs data in very different forms (so, at best, the callback would have a return type like `-callback get_telemetry_info() -> map().`).  

With the approach taken in the last few PRs suggested by @zmstone (i.e., creating functions in each application that exports data relevant to telemetry), do you think that defining the behaviour would be much different (except for the auto-discover, which sounds like a cool idea :sunglasses: )?

**Comment by thalesmg** [comment]
@savonarola I'll merge this one now, and if we decide to apply that refactoring then I'll follow that up in new PRs. :smile_cat:

**Comment by savonarola** [comment]
@thalesmg my comment was more like some general thoughts for consideration

**Pull request merged**